pkg/controllers/podintent_reconciler.go

```diff
@@ -146,54 +143,59 @@ func ResolveConventions() reconcilers.SubReconciler {
+	return &reconcilers.SyncReconciler[*conventionsv1alpha1.PodIntent]{
+		Name: "BuildRegistryConfigReconciler",
+		SyncWithResult: func(ctx context.Context, parent *conventionsv1alpha1.PodIntent) (ctrl.Result, error) {
 			log := logr.FromContextOrDiscard(ctx)
 			c := reconcilers.RetrieveConfigOrDie(ctx)
 			if rc.Client == nil {
 				return ctrl.Result{}, fmt.Errorf("kubernetes client is not set")
 			}
 			conditionManager := parent.GetConditionSet().Manage(&parent.Status)
-			parentNamespacedName := types.NamespacedName{Namespace: parent.Namespace, Name: parent.Name}
 
-			var imagePullSecrets []string
+			// create list of image pull secrets from parent spec
+			var pullSecretNamesList []string
 			for _, ips := range parent.Spec.ImagePullSecrets {
-				imagePullSecrets = append(imagePullSecrets, ips.Name)
-				// track ref for updates
-				key := tracker.NewKey(secretGVK, types.NamespacedName{Namespace: parent.Namespace, Name: ips.Name})
-				c.Tracker.Track(ctx, key, parentNamespacedName)
+				pullSecretNamesList = append(pullSecretNamesList, ips.Name)
 			}
 
 			serviceAccountName := parent.Spec.ServiceAccountName
 
+			// initalise a new k8schain using the parent service account and pull secrets
 			kc, err := k8schain.New(ctx, rc.Client, k8schain.Options{
 				Namespace:          parent.Namespace,
 				ServiceAccountName: serviceAccountName,
-				ImagePullSecrets:   imagePullSecrets,
+				ImagePullSecrets:   pullSecretNamesList,
 			})
+
 			if err != nil {
 				conditionManager.MarkFalse(conventionsv1alpha1.PodIntentConditionConventionsApplied, "ImageResolutionFailed", "failed to authenticate: %v", err.Error())
 				log.Error(err, "fetching authentication for Images failed")
 				return ctrl.Result{}, nil
 			}
 
 			serviceAccountNamespacedName := types.NamespacedName{Namespace: parent.Namespace, Name: serviceAccountName}
-			sa := &corev1.ServiceAccount{}
-			if err = c.TrackAndGet(ctx, serviceAccountNamespacedName, sa); err != nil {
+			serviceAccount := &corev1.ServiceAccount{}
+			if err = c.TrackAndGet(ctx, serviceAccountNamespacedName, serviceAccount); err != nil {
 				log.Error(err, "fetching serviceAccount failed")
 				// should not happen mostly.
+
 				conditionManager.MarkFalse(conventionsv1alpha1.PodIntentConditionConventionsApplied, "ImageResolutionFailed", fmt.Sprintf("failed to authenticate: %v", err.Error()))
 				return ctrl.Result{}, nil
 			}
 
-			for _, secretReference := range sa.ImagePullSecrets {
-				// track ref for updates
-				key := tracker.NewKey(secretGVK, types.NamespacedName{Namespace: parent.Namespace, Name: secretReference.Name})
-				c.Tracker.Track(ctx, key, parentNamespacedName)
+			// lookup each individual image pull secret from the service account
+			for _, secretReference := range serviceAccount.ImagePullSecrets {
+				pullSecretObj := corev1.Secret{}
+				err := c.TrackAndGet(ctx, types.NamespacedName{Namespace: parent.Namespace, Name: secretReference.Name}, &pullSecretObj)
```

Comment:
there's no need to load a resource just to track it. Use https://pkg.go.dev/github.com/vmware-labs/reconciler-runtime@v0.12.0/tracker#Tracker.TrackReference